### ISSUE TITLE
Force NCHW format

### ISSUE THREAD

**Issue by lukemurray** [onnx/onnx-tensorflow#31]
Hi there, this seems to add a bunch of transpose ops into the graph, can we force it to build the graph in NCHW format?

**Comment by tjingrant** [comment]
Are you building your graph on a CPU? Transpositions only occur when NHWC format is needed for computation (typically when you have no GPU). Fundamentally if you only have CPUs, there's no way you can execute that graph because TF does not have full support for NCHW format on CPU. (Although you may still want to build the graph that way despite the fact that you cannot run it, but I suspect this is not your concern.)

**Comment by lukemurray** [comment]
I guess I want a typical tf model format actually. It is going into another third-party tool that doesn't currently support the transposes. I'm changing it so it transposes the raw data and therefore creates a cleaner tf graph without any transforms. Which this third-party tool will support.

Any reason you didn't implement it that way? I guess in general transpose is not expensive.

________________________________

**Comment by tjingrant** [comment]
This is because we assume that the user will run the graph in the same machine as he/she converts the graph, which is the case for us, but may not be the case for others. We may support exporting GPU-compatible models on a CPU-only machine in the future. I'll take note and see what I can do.

**Comment by lukemurray** [comment]
Yep, that makes sense. I think I can get this working for my case. But I would say the changes are nice enough for a PR 😁

________________________________

**Comment by lukemurray** [comment]
Actually if I understand you correctly. I can load the onnx model and save the tf model on a computer with cuda and there will be no transforms in the tf graph?

________________________________

**Comment by tjingrant** [comment]
Yes, we basically ask TF whether there's a CUDA device and if there's not, we perform transpositions. The condition is checked here:
https://github.com/onnx/onnx-tensorflow/blob/ca5c120d12282f11152ebc3ef29141f369953f66/onnx_tf/backend.py#L990

And the decision to prepend/append transpositions to an op is made in every op that concerns data format, for instance, the pooling op has the following logic:
https://github.com/onnx/onnx-tensorflow/blob/ca5c120d12282f11152ebc3ef29141f369953f66/onnx_tf/backend.py#L543

And contributions are always welcome!

**Comment by LucasMahieu** [comment]
I have though a lot about this problem. Here is a proposal of how I think things have to be done to be optimal.

I will give an example to explain my idea:
I suggest that the user have the choice of two things:

    - The device on which he wants to run the TF graph
          (the source graph for front-end or the destination graph for back-end) (GPU or CPU)
    - The data format that he will put as input of the TF graph 
          (the source graph for front-end or the destination graph for back-end)  (NCHW or NHWC)

Back-end
--- 

Now, assume that the ONNX graph is : 
**-> MaxPool(k, s, p) -> Concat(axis = 1) ->**
k, p, s is kernel, stride and pad for H and W dimension it will not be impact by the thereafter explanation.
This example illustrate how to deal with ops that have specific constrains on data format and others that don't have.

Assuming that the input given in the ONNX graph is always in 'NCHW' format.
So if user makes choice to give 'NHWC' data to the TF graph, he has to transpose all the inputs taken from ONNX.

According to the 2 choices of the user, if he wants to convert this ONNX graph to Tensorflow Graph, we have 4 different cases to support.

**ONNX -> Tensorflow: Each cell of the Table is the TF graph generated**

| device\input data <br>of TF graph	| NCHW                      | NHWC 	|
|-------------------------                |------                          |-----               |
| GPU                                       | maxpool<br>concat(axis=1) | transpose(0,3,1,2)<br>maxpool<br>transpose(0,2,3,1)<br>concat(axis=3)|
| CPU                                        |transpose(0,2,3,1)<br>maxpool<br>transpose(0,3,1,2)<br>concat(axis=1) | maxpool<br>concat(axis=3) |

Front-end
--- 

For the front-end, is the same, the user has to said if the TF graph is made for CPU or GPU and if the tensors of the graph are in format NCHW or NHWC.

Now, assume that the TF graph is : 
**-> MaxPool(k, s, p, d) -> Concat(axis=channel) ->**
k, p, s is kernel, stride and padding type for all dimensions. d in the data_format of the op's input.
In TF, k=[k0, k1, k2, k3] ; s=[s0, s1, s2, s3] ; one value for each axis, so they will be impacted by user's choices.

**Tensorflow -> ONNX: Each cell of the Table is the ONNX graph generated**

| device\ data format <br>of TF graph | NCHW                      | NHWC 	|
|-------------------------                |------                          |-----               |
| GPU                                       | maxpool((k2,k3),(s2,s3))<br>concat(axis=1) | \* note 1 |
| CPU                                        | \* note 2 | maxpool((k1,k2),(s1,s2))<br>concat(axis=1) |

- note 1 : This case is special. If the TF graph is just like this : 
*-> MaxPool(k, s, p, 'NHWC') -> Concat(axis=3) ->* 
It would not be executable on CPU with TF. But if  the user takes care of that and use transposed just before and after the MaxPool(k, s, p, 'NHWC') (or if the tf graph was generated by our backend for example) the graph would be correct.
And so, the onnx-tf front-end should parse and transform : 

    **This tf graph part:**
*-> transpose(0,3,1,2) -> MaxPool((k0,k1,k2,k3), (s0,s1,s2,s3), p, 'NHWC') -> transpose(0,2,3,1) -> Concat(axis=3) ->*
    **Into this onnx graph part**
*-> MaxPool((k1,k2), (s1,s2), p) -> Concat(axis=1) ->*


- note 2 : This case is special. If the TF graph is just like this : 
*-> MaxPool(k, s, p, 'NCHW') -> Concat(axis=1) ->* 
It would not be executable on CPU with TF. But if  the user takes care of that and use transposed just before and after the MaxPool(k, s, p, 'NHWC') (or if the tf graph was generated by our backend for example) the graph would be correct.
And so, the onnx-tf front-end should parse and transform : 

    **This tf graph part:**
*-> transpose(0,2,3,1) -> MaxPool((k0,k1,k2,k3), (s0,s1,s2,s3), p, 'NHWC') -> transpose(0,3,1,2) -> Concat(axis=1) ->*
    **Into this onnx graph part**
*-> MaxPool((k2,k3), (s2,s3), p) -> Concat(axis=1) ->*


To conclude, we have to add the 2 possible choices for the user when converting a graph.
And we have to parse graph in non one after one order, but in a more intelligent way according to the user choices in order to support special cases "note 1" and "note 2". This could be made by two different methods : 
    - Adding a transpose op around max_pool, then remove 2 opposite successive transpose ops from the final graph.
    - Identify a pattern "transpose + max_pool + transpose" and generate only a max_pool with the rights parameters

**Comment by tjingrant** [comment]
@LucasMahieu are you suggesting that the user can specify a **default** data format for the entire graph? Or just a data format for the **input** for the entire graph (and thus after the first op, we then default to NCHW)?

**Comment by lukemurray** [comment]
Can you use NCHW the whole way through the graph?

TF gives this error on a MaxPool
>InvalidArgumentError (see above for traceback): Default MaxPoolingOp only supports NHWC.

It seems NHWC works through the whole graph in TF. 

I think allowing to export without the in-place transposes would be good. Well it would certainly suit me, and it removes operations from the graph if you are running it heavily.

**Comment by LucasMahieu** [comment]
@lukemurray You are right, it is exactly the problem. Please look at the "*note 1 and 2" of the https://github.com/onnx/onnx-tensorflow/issues/31#issuecomment-352019711. It is exactly the case you are talking about.

I really think that exporting a tf graph to ONNX then exporting back to tf should give EXACTLY the same tf graph. Without any additional transpose, etc.

**Comment by LucasMahieu** [comment]
Please look at https://github.com/onnx/onnx/issues/369#
This should be really interesting for us. Please, help to convince onnx community to support data format natively.

This should solve all our issues about data format conversion.

**Comment by LucasMahieu** [comment]
@tjingrant I am suggesting that the user can specify a default data format for the entire graph. 
So, for all operation like 'concat', the converter has just to look at the data format of the graph that the user had specified, and to the device the used specified too. According to this 2 variables, the converter can output the right thing. see my two table for the output to produce by the converter according to these two variables.

**Comment by lukemurray** [comment]
This is weird. I have changed the code so that that the TF graph never does in-graph transpose, it assumes the data coming in the input is `NHWC` and I transpose the weights before creating the `tf.constant`s.

Problem is, the graph produces completely different results. I'm been through it a million times. But I must be missing something.

Even if I keep the in-graph transposes but pre-transpose the weights before creating the `tf.constant`s it works (i.e. removing 1 transpose per conv layer in the graph), so I don't think it is the weights being incorrect.

Just thought I'd see if you had any ideas of what I might be missing?

Thanks,

**Comment by LucasMahieu** [comment]
Need to modify the proposal according to thins : https://github.com/onnx/onnx-tensorflow/pull/43

**Comment by tjingrant** [comment]
@LucasMahieu so basically they decided that we keep our transpositions, but instead their graph optimizers are going to take the redundant transpositions away. I do not think this is a good idea tho... IBM Research people are in contact with the ONNX people off Github and will let you know if anything meaningful comes out of the discussion.

**Comment by dmoham1476** [comment]
Could you please suggest how I can get rid of the transposes in the tensorflow graph generated from onnx on a CPU device?

**Comment by fumihwh** [comment]
@dmoham1476 
There is no way to avoid this currently.
In tensorflow, you can only use NHWC for most ops on CPU, which is always NCHW in onnx.
And transpose op is also used for transposing kernel etc for compatibility. They have different format in onnx and tensorflow.
Without transpose, your graph might not work properly.

**Comment by dmoham1476** [comment]
Hi, I ran an inference with tensorflow+Intel MKL and it seems like MKL is able to do the layout conversions on the tensorflow graph like the GPU kernels. This presentation for reference - https://pdfs.semanticscholar.org/presentation/3805/d3288ff3a435172c4d95e76198a1bbffc725.pdf (Kindly look at slide 13). So it would be great if there is a way to eliminate the transposes in the CPU graph when we  convert from tensorflow ->onnx and vice versa and leave it to the HW library.

**Comment by fumihwh** [comment]
@dmoham1476 As I know, ONNX has optimizer for continuous transpose. You can try that.

**Comment by chinhuang007** [comment]
No activities for more than a year. Please reopen if anyone wants to continue the discussions.

**Comment by AlexeyAB** [comment]
@chinhuang007 
Do I understand correctly that conversion ONNX(NCHW) -> PB(NHWC) is not currently supported?

PR was merged https://github.com/onnx/onnx-tensorflow/pull/39 but then it was reverted https://github.com/onnx/onnx-tensorflow/pull/43

**Comment by chinhuang007** [comment]
Conversion ONNX(NCHW) -> PB(NHWC) is currently supported if the system is CPU only. The user input of device type to prepare API will be worked on shortly.

**Comment by Vozf** [comment]
In 1.7.0 we have `device=’CPU’` or  `device=’CUDA’`. But is there a way to fully convert nchw to nhwc and avoid all the transpose ops by fully integrating nhwc for the whole data flow? It seems like this was done in the above mentioned reverted PR. What is the current state of such conversion?

**Comment by chinhuang007** [comment]
Currently there is no implementation to fully convert nchw to nhwc for the entire model and avoid transpose ops. We are working with the ONNX core community to see if an optional NHWC data format can be added to the spec so backend converters like onnx-tf can easily support channel last format.

**Comment by Vozf** [comment]
For anyone wondering a way for full nhwc. You can achieve that with conversion to keras where converter allows the full ordering change
https://github.com/waltYeh/onnx2keras
